File: lib/services/notification_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class NotificationService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  /// Generic notification creator
  Future<void> sendNotification({
    required String userId,
    required String title,
    required String message,
    required String type,
    required String lotteryId,
    String? action,
    Map<String, dynamic>? extraData,
  }) async {
    await _firestore
        .collection('users')
        .doc(userId)
        .collection('notifications')
        .add({
      'title': title,
      'message': message,
      'type': type,
      'lotteryId': lotteryId,
      'action': action,
      'isRead': false,
      'createdAt': FieldValue.serverTimestamp(),
      'extraData': extraData ?? {},
    });
  }

  /// Creator notification when lottery sells out
  Future<void> notifyLotterySoldOut({
    required String creatorId,
    required String creatorName,
    required String lotteryId,
    required String lotteryTitle,
    required int totalTickets,
  }) async {
    await sendNotification(
      userId: creatorId,
      title: "🎉 Your lottery is sold out!",
      message:
          "Hi $creatorName,\n\nYour lottery \"$lotteryTitle\" has officially sold all $totalTickets tickets.\n\nThe draw is now ready.",
      type: "sold_out",
      lotteryId: lotteryId,
      action: "view_lottery",
      extraData: {
        'totalTickets': totalTickets,
        'creatorName': creatorName,
      },
    );
  }

  /// Creator notification after draw (UPDATED with more details)
  Future<void> notifyDrawCompleted({
    required String creatorId,
    required String creatorName,
    required String lotteryId,
    required String lotteryTitle,
    int? winnersCount,
    int? totalParticipants,
    List<String>? winnerIds,
    List<int>? winningNumbers,
  }) async {
    String message;
    
    if (winnersCount != null && totalParticipants != null) {
      message = "Hi $creatorName,\n\nThe draw for \"$lotteryTitle\" has finished!\n\n"
          "🎯 $winnersCount winner(s) selected from $totalParticipants participants.\n\n"
          "Tap to view the winners.";
    } else {
      message = "Hi $creatorName,\n\nThe draw for \"$lotteryTitle\" has finished.\n\n"
          "Tap to view the winners.";
    }

    await sendNotification(
      userId: creatorId,
      title: "🏆 Draw Completed!",
      message: message,
      type: "draw_complete",
      lotteryId: lotteryId,
      action: "view_winners",
      extraData: {
        'winnersCount': winnersCount,
        'totalParticipants': totalParticipants,
        'winnerIds': winnerIds,
        'winningNumbers': winningNumbers,
        'creatorName': creatorName,
      },
    );
  }

  /// 🆕 Creator reminder for upcoming draw
  Future<void> notifyCreatorUpcomingDraw({
    required String creatorId,
    required String creatorName,
    required String lotteryId,
    required String lotteryTitle,
    required DateTime drawTime,
    int? ticketsSold,
    int? totalTickets,
  }) async {
    final timeRemaining = drawTime.difference(DateTime.now());
    final hoursRemaining = timeRemaining.inHours;
    final minutesRemaining = timeRemaining.inMinutes;

    String timeMessage;
    if (hoursRemaining > 0) {
      timeMessage = '$hoursRemaining hours';
    } else if (minutesRemaining > 0) {
      timeMessage = '$minutesRemaining minutes';
    } else {
      timeMessage = 'very soon';
    }

    String ticketStatus = '';
    if (ticketsSold != null && totalTickets != null) {
      ticketStatus = '\n📊 $ticketsSold out of $totalTickets tickets sold.';
    }

    await sendNotification(
      userId: creatorId,
      title: "⏰ Draw Reminder",
      message: "Hi $creatorName,\n\nThe draw for \"$lotteryTitle\" is scheduled in $timeMessage!$ticketStatus\n\n"
          "Make sure everything is ready.",
      type: "draw_reminder",
      lotteryId: lotteryId,
      action: "view_lottery",
      extraData: {
        'drawTime': drawTime.toIso8601String(),
        'hoursRemaining': hoursRemaining,
        'minutesRemaining': minutesRemaining,
        'ticketsSold': ticketsSold,
        'totalTickets': totalTickets,
        'creatorName': creatorName,
      },
    );
  }

  /// 🆕 Notification for low ticket sales (when draw time is near)
  Future<void> notifyCreatorLowTickets({
    required String creatorId,
    required String creatorName,
    required String lotteryId,
    required String lotteryTitle,
    required int ticketsSold,
    required int totalTickets,
    required DateTime drawTime,
  }) async {
    final remainingTickets = totalTickets - ticketsSold;
    final drawTimeFormatted = '${drawTime.hour.toString().padLeft(2, '0')}:${drawTime.minute.toString().padLeft(2, '0')}';

    await sendNotification(
      userId: creatorId,
      title: "⚠️ Low Ticket Sales Alert",
      message: "Hi $creatorName,\n\nYour lottery \"$lotteryTitle\" has only $remainingTickets tickets remaining.\n\n"
          "The draw is scheduled for $drawTimeFormatted. Consider promoting it!",
      type: "low_tickets",
      lotteryId: lotteryId,
      action: "promote_lottery",
      extraData: {
        'ticketsSold': ticketsSold,
        'totalTickets': totalTickets,
        'remainingTickets': remainingTickets,
        'drawTime': drawTime.toIso8601String(),
        'creatorName': creatorName,
      },
    );
  }

  /// Winner notification (UPDATED with ticket number)
  Future<void> notifyWinner({
    required String userId,
    required String lotteryId,
    required String lotteryTitle,
    int? ticketNumber,
  }) async {
    String message;
    if (ticketNumber != null) {
      message = "🎉 Congratulations!\n\nYou won \"$lotteryTitle\" with ticket #$ticketNumber!\n\n"
          "Tap here to claim your prize.";
    } else {
      message = "🎉 Congratulations!\n\nYou won \"$lotteryTitle\"!\n\n"
          "Tap here to claim your prize.";
    }

    await sendNotification(
      userId: userId,
      title: "🎉 Congratulations!",
      message: message,
      type: "winner",
      lotteryId: lotteryId,
      action: "claim_prize",
      extraData: {
        'ticketNumber': ticketNumber,
        'lotteryTitle': lotteryTitle,
      },
    );
  }

  /// Loser notification (UPDATED with more details)
  Future<void> notifyLoser({
    required String userId,
    required String lotteryId,
    required String lotteryTitle,
    int? ticketNumber,
  }) async {
    String message;
    if (ticketNumber != null) {
      message = "🍀 Better Luck Next Time\n\nThe draw for \"$lotteryTitle\" has ended.\n\n"
          "Your ticket #$ticketNumber wasn't selected this time.\n\n"
          "Better luck next time!";
    } else {
      message = "🍀 Better Luck Next Time\n\nThe draw for \"$lotteryTitle\" has ended.\n\n"
          "Unfortunately your ticket wasn't selected this time.\n\n"
          "Better luck next time!";
    }

    await sendNotification(
      userId: userId,
      title: "🍀 Better Luck Next Time",
      message: message,
      type: "loser",
      lotteryId: lotteryId,
      action: "buy_again",
      extraData: {
        'ticketNumber': ticketNumber,
        'lotteryTitle': lotteryTitle,
      },
    );
  }

  /// Mark notification as read
  Future<void> markAsRead({
    required String userId,
    required String notificationId,
  }) async {
    await _firestore
        .collection('users')
        .doc(userId)
        .collection('notifications')
        .doc(notificationId)
        .update({
      'isRead': true,
    });
  }

  /// 🆕 Mark all notifications as read
  Future<void> markAllAsRead({
    required String userId,
  }) async {
    final notifications = await _firestore
        .collection('users')
        .doc(userId)
        .collection('notifications')
        .where('isRead', isEqualTo: false)
        .get();

    final batch = _firestore.batch();
    for (final doc in notifications.docs) {
      batch.update(doc.reference, {'isRead': true});
    }
    await batch.commit();
  }

  /// 🆕 Get unread notification count (FIXED)
  Future<int> getUnreadNotificationCount({
    required String userId,
  }) async {
    try {
      // Try using count() first (newer Firebase version)
      final snapshot = await _firestore
          .collection('users')
          .doc(userId)
          .collection('notifications')
          .where('isRead', isEqualTo: false)
          .count()
          .get();
      
      return snapshot.count ?? 0;
    } catch (e) {
      // Fallback to fetching all docs (older Firebase version)
      final snapshot = await _firestore
          .collection('users')
          .doc(userId)
          .collection('notifications')
          .where('isRead', isEqualTo: false)
          .get();
      
      return snapshot.docs.length;
    }
  }

  /// 🆕 Delete old notifications
  Future<void> deleteOldNotifications({
    required String userId,
    int daysOld = 30,
  }) async {
    final cutoffDate = DateTime.now().subtract(Duration(days: daysOld));
    
    final oldNotifications = await _firestore
        .collection('users')
        .doc(userId)
        .collection('notifications')
        .where('createdAt', isLessThan: Timestamp.fromDate(cutoffDate))
        .get();

    final batch = _firestore.batch();
    for (final doc in oldNotifications.docs) {
      batch.delete(doc.reference);
    }
    await batch.commit();
  }
}

File: backend/src/services/notification.service.ts
import { db } from "../config/firebase";

export type NotificationType =
  | "LOTTERY_STARTED"
  | "LOTTERY_COMPLETED"
  | "LOTTERY_WON"
  | "LOTTERY_LOST"
  | "LOTTERY_SOLD_OUT";

export class NotificationService {
  /**
   * Create a notification for a user.
   */
  async createNotification({
    userId,
    title,
    message,
    type,
    lotteryId,
    lotteryTitle,
    data = {},
  }: {
    userId: string;
    title: string;
    message: string;
    type: NotificationType;
    lotteryId: string;
    lotteryTitle: string;
    data?: Record<string, any>;
  }) {
    if (!userId) {
      console.warn(
        "⚠️ Notification skipped: missing userId",
      );
      return;
    }

    await db.collection("notifications").add({
      userId,

      title,

      message,

      type,

      lotteryId,

      lotteryTitle,

      isRead: false,

      data,

      createdAt: new Date(),
    });
  }

  /**
   * Notify the creator that their lottery
   * has started drawing.
   */
  async notifyLotteryStarted({
    creatorId,
    lotteryId,
    lotteryTitle,
  }: {
    creatorId: string;
    lotteryId: string;
    lotteryTitle: string;
  }) {
    await this.createNotification({
      userId: creatorId,

      title: "🎰 Lottery Draw Started",

      message:
        `"${lotteryTitle}" is now being drawn. ` +
        "Good luck to all participants!",

      type: "LOTTERY_STARTED",

      lotteryId,

      lotteryTitle,

      data: {
        status: "DRAWING",
      },
    });
  }

  /**
   * Notify the creator that the lottery
   * has been completed.
   */
  async notifyLotteryCompleted({
    creatorId,
    lotteryId,
    lotteryTitle,
    winnerCount,
    winningTicketNumbers,
    winnerIds,
  }: {
    creatorId: string;
    lotteryId: string;
    lotteryTitle: string;
    winnerCount: number;
    winningTicketNumbers: number[];
    winnerIds: string[];
  }) {
    await this.createNotification({
      userId: creatorId,

      title: "🏆 Lottery Completed",

      message:
        winnerCount === 1
          ? `"${lotteryTitle}" has been completed and a winner has been selected!`
          : `"${lotteryTitle}" has been completed and ${winnerCount} winners have been selected!`,

      type: "LOTTERY_COMPLETED",

      lotteryId,

      lotteryTitle,

      data: {
        winnerCount,
        winnerIds,
        winningTicketNumbers,
        status: "COMPLETED",
      },
    });
  }

  /**
   * Notify a user that they WON.
   */
  async notifyWinner({
    userId,
    lotteryId,
    lotteryTitle,
    ticketNumber,
    jackpot,
  }: {
    userId: string;
    lotteryId: string;
    lotteryTitle: string;
    ticketNumber: number;
    jackpot: number;
  }) {
    await this.createNotification({
      userId,

      title: "🎉 YOU WON!",

      message:
        `Congratulations! Your ticket #${ticketNumber} ` +
        `won "${lotteryTitle}"!`,

      type: "LOTTERY_WON",

      lotteryId,

      lotteryTitle,

      data: {
        ticketNumber,
        jackpot,
        result: "WINNER",
      },
    });
  }

  /**
   * Notify a user that they did NOT win.
   */
  async notifyLoser({
    userId,
    lotteryId,
    lotteryTitle,
    ticketNumber,
  }: {
    userId: string;
    lotteryId: string;
    lotteryTitle: string;
    ticketNumber: number;
  }) {
    await this.createNotification({
      userId,

      title: "Lottery Result",

      message:
        `Ticket #${ticketNumber} did not win ` +
        `"${lotteryTitle}". Better luck next time! 💜`,

      type: "LOTTERY_LOST",

      lotteryId,

      lotteryTitle,

      data: {
        ticketNumber,
        result: "NOT_WINNER",
      },
    });
  }

  /**
   * Notify the creator that all tickets
   * have been sold.
   */
  async notifyLotterySoldOut({
    creatorId,
    lotteryId,
    lotteryTitle,
    totalTickets,
  }: {
    creatorId: string;
    lotteryId: string;
    lotteryTitle: string;
    totalTickets: number;
  }) {
    await this.createNotification({
      userId: creatorId,

      title: "🎟️ Lottery Sold Out",

      message:
        `"${lotteryTitle}" has sold all ` +
        `${totalTickets} tickets!`,

      type: "LOTTERY_SOLD_OUT",

      lotteryId,

      lotteryTitle,

      data: {
        totalTickets,
        status: "SOLD_OUT",
      },
    });
  }

  /**
   * Get all notifications for a user.
   */
  async getUserNotifications(
    userId: string,
  ) {
    const snapshot = await db
      .collection("notifications")
      .where("userId", "==", userId)
      .get();

    const notifications =
      snapshot.docs.map((doc) => ({
        id: doc.id,
        ...doc.data(),
      }));

    /**
     * Newest notifications first.
     */
    notifications.sort(
      (a: any, b: any) => {
        const aTime =
          a.createdAt?.toMillis?.() ?? 0;

        const bTime =
          b.createdAt?.toMillis?.() ?? 0;

        return bTime - aTime;
      },
    );

    return notifications;
  }

  /**
   * Mark one notification as read.
   */
  async markAsRead(
    notificationId: string,
  ) {
    await db
      .collection("notifications")
      .doc(notificationId)
      .update({
        isRead: true,
      });
  }

  /**
   * Mark all notifications belonging
   * to a user as read.
   */
  async markAllAsRead(
    userId: string,
  ) {
    const snapshot = await db
      .collection("notifications")
      .where("userId", "==", userId)
      .where("isRead", "==", false)
      .get();

    if (snapshot.empty) {
      return;
    }

    const batch = db.batch();

    for (
      const doc of snapshot.docs
    ) {
      batch.update(doc.ref, {
        isRead: true,
      });
    }

    await batch.commit();
  }
}